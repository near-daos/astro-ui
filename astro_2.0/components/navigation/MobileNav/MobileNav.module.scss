@import 'styles/mixins';

.root {
  align-items: center;
  background-color: var(--color-white);
  border-radius: 8px 8px 0 0;
  bottom: 0;
  box-shadow: 3px 2px 24px var(--color-black-opacity-1);
  display: flex;
  height: 64px;
  justify-content: space-around;
  position: fixed;
  width: 100%;
  z-index: 1000;

  @include desktop {
    display: none;
  }
}
