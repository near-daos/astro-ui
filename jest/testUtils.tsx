import {
  RenderResult,
  render as reactTestingLibRender,
} from '@testing-library/react';
import { ReactNode } from 'react';
import { NextRouter } from 'next/router';
import { RouterContext } from 'next/dist/shared/lib/router-context';

function createMockRouter(router?: Partial<NextRouter>): NextRouter {
  return {
    basePath: '',
    pathname: '/',
    route: '/',
    query: {},
    asPath: '/',
    back: jest.fn(),
    beforePopState: jest.fn(),
    prefetch: jest.fn(),
    push: jest.fn(),
    reload: jest.fn(),
    replace: jest.fn(),
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
    isFallback: false,
    isLocaleDomain: false,
    isReady: true,
    defaultLocale: 'en',
    domainLocales: [],
    isPreview: false,
    ...router,
  };
}

export function render(
  component: ReactNode,
  router?: Partial<NextRouter>
): RenderResult {
  return reactTestingLibRender(
    <RouterContext.Provider value={createMockRouter(router)}>
      {component}
    </RouterContext.Provider>
  );
}
