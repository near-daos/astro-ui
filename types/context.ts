import { ProposalFeedItem, ProposalType } from 'types/proposal';
import { DAO } from 'types/dao';

export type UserPermissions = {
  isCanCreateProposals: boolean;
  isCanCreatePolicyProposals: boolean;
  allowedProposalsToCreate: ProposalPermissions;
  allowedProposalsToVote: ProposalPermissions;
};

export type ProposalPermissions = {
  [ProposalType.ChangeConfig]: boolean;
  [ProposalType.ChangePolicy]: boolean;
  [ProposalType.AddBounty]: boolean;
  [ProposalType.BountyDone]: boolean;
  [ProposalType.Transfer]: boolean;
  [ProposalType.Vote]: boolean;
  [ProposalType.RemoveMemberFromRole]: boolean;
  [ProposalType.AddMemberToRole]: boolean;
  [ProposalType.AddMemberToRole]: boolean;
  [ProposalType.SetStakingContract]: boolean;
  [ProposalType.UpgradeSelf]: boolean;
  [ProposalType.UpgradeRemote]: boolean;
  [ProposalType.FunctionCall]: boolean;
};

export type DaoContext = {
  userPermissions: UserPermissions;
  dao: DAO;
  policyAffectsProposals: ProposalFeedItem[];
};
