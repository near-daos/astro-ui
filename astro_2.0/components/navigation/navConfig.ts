import {
  ALL_DAOS_URL,
  ALL_FEED_URL,
  CFC_LIBRARY,
  CREATE_DAO_URL,
  DISCOVER,
  MY_DAOS_URL,
  MY_FEED_URL,
} from 'constants/routing';

import { NavItemProps } from './types';

export const DISCOVER_NAV_CONFIG: NavItemProps = {
  icon: 'aDiscover',
  hoverIcon: 'aDiscover',
  href: DISCOVER,
  label: 'Discovery',
};

export const CFC_LIBRARY_NAV_CONFIG: NavItemProps = {
  icon: 'proposalFunctionCall',
  hoverIcon: 'proposalFunctionCall',
  href: CFC_LIBRARY,
  label: 'Actions Library',
};

export const ALL_DAOS_NAV_CONFIG: NavItemProps = {
  icon: 'aAllDaos',
  hoverIcon: 'aAllDaosHover',
  href: ALL_DAOS_URL,
  label: 'All DAOs',
};

export const ASTRO_FEED_NAV_CONFIG: NavItemProps = {
  icon: 'aAstroFeed',
  hoverIcon: 'aAstroFeedHover',
  href: ALL_FEED_URL,
  label: 'Global Feed',
};

export const MY_DAOS_NAV_CONFIG: NavItemProps = {
  icon: 'aMyDaos',
  hoverIcon: 'aMyDaosHover',
  href: MY_DAOS_URL,
  label: '  My DAOs',
};

export const MY_FEED_NAV_CONFIG: NavItemProps = {
  icon: 'aMyFeed',
  hoverIcon: 'aMyFeedHover',
  href: MY_FEED_URL,
  label: 'My Feed',
};

export const CREATE_DAO_NAV_CONFIG: NavItemProps = {
  icon: 'aCreateDao',
  hoverIcon: 'aCreateDaoHover',
  href: CREATE_DAO_URL,
  label: 'Create DAO',
  authRequired: true,
};
