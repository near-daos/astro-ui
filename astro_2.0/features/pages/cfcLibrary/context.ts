import { createContext, useContext } from 'react';
import { DaoFeedItem } from 'types/dao';

interface Props {
  accountDaos: DaoFeedItem[];
  accountId: string;
  onUpdate: (templateId: string, daosCount: number) => void;
}

export const CfcLibraryContext = createContext<Props>({
  accountDaos: [],
  accountId: '',
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  onUpdate: () => {},
});

export const useCfcValues = (): Props => useContext<Props>(CfcLibraryContext);
