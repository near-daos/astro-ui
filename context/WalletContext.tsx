import { WalletType } from 'types/config';
import {
  WalletMeta,
  WalletService,
} from 'services/sputnik/SputnikNearService/walletServices/types';
import { createContext, FC, useCallback, useContext, useMemo } from 'react';
import { useBoolean } from 'react-use';

import { configService } from 'services/ConfigService';
import { CookieService } from 'services/CookieService';
import { ACCOUNT_COOKIE } from 'constants/cookies';
import { SputnikNearService } from 'services/sputnik';
import { useRouter } from 'next/router';
import { ConnectingWalletModal } from 'astro_2.0/features/Auth/components/ConnectingWalletModal';
import {
  PkAndSignature,
  useAvailableAccounts,
  useWallet,
  usePkAndSignature,
} from 'context/WalletContextHooks';
import { GA_EVENTS, sendGAEvent } from 'utils/ga';

export interface WalletContext {
  availableWallets: WalletMeta[];
  currentWallet: WalletType | null;
  accountId: string;
  connectingToWallet: boolean;
  login: (walletType: WalletType) => Promise<void>;
  logout: () => Promise<void>;
  switchAccount: (walletType: WalletType, accountId: string) => void;
  switchWallet: (walletType: WalletType) => void;
  availableAccounts: string[];
  pkAndSignature: PkAndSignature | null;
  // todo get rid of
  nearService: SputnikNearService | null;
}

const WalletContext = createContext<WalletContext>({} as WalletContext);

export const WrappedWalletContext: FC = ({ children }) => {
  const [
    currentWallet,
    availableWallets,
    getWallet,
    setWallet,
    removePersistedWallet,
  ] = useWallet();
  const availableAccounts = useAvailableAccounts(currentWallet);
  const pkAndSignature = usePkAndSignature(currentWallet);
  const { nearConfig } = configService.get();

  const [connectingToWallet, toggleConnection] = useBoolean(false);

  const router = useRouter();

  const signIn = useCallback(
    async (wallet: WalletService, contractName: string) => {
      toggleConnection(true);

      if (!wallet.isSignedIn()) {
        await wallet.signIn(contractName);
      }

      setWallet(wallet);

      toggleConnection(false);

      sendGAEvent({
        name: GA_EVENTS.SIGN_IN,
        accountId: wallet.getAccountId(),
      });
    },
    [setWallet, toggleConnection]
  );

  const login = useCallback(
    async (walletType: WalletType) => {
      const wallet = getWallet(walletType);

      if (!wallet) {
        return;
      }

      signIn(wallet, nearConfig.contractName);
    },
    [getWallet, nearConfig.contractName, signIn]
  );

  const logout = useCallback(async () => {
    sendGAEvent({
      name: GA_EVENTS.SIGN_OUT,
      accountId: currentWallet?.getAccountId() ?? '',
    });

    CookieService.remove(ACCOUNT_COOKIE);
    removePersistedWallet();
    availableWallets.forEach(wallet => wallet.logout());
    router.reload();
  }, [availableWallets, currentWallet, removePersistedWallet, router]);

  const switchAccount = useCallback(
    async (walletType: WalletType, accountId: string) => {
      // Currently its NEAR wallet specific only
      if (walletType !== WalletType.NEAR) {
        return;
      }

      const nearWallet = getWallet(WalletType.NEAR);

      if (!nearWallet) {
        return;
      }

      // eslint-disable-next-line no-underscore-dangle
      const keypair = await nearWallet
        .getKeyStore()
        .getKey(nearConfig.networkId, accountId);

      if (!keypair) {
        return;
      }

      const authData = {
        accountId,
        allKeys: [keypair.getPublicKey().toString()],
      };

      // new wallet instance will take the new auth_key and will reinit the account
      window.localStorage.setItem(
        'sputnik_wallet_auth_key',
        JSON.stringify(authData)
      );

      CookieService.set(ACCOUNT_COOKIE, accountId, {
        path: '/',
      });

      sendGAEvent({
        name: GA_EVENTS.SWITCH_ACCOUNT,
        accountId,
      });

      router.reload();
    },
    [getWallet, nearConfig.networkId, router]
  );

  const switchWallet = useCallback(
    async (walletType: WalletType) => {
      const selectedWallet = getWallet(walletType);

      if (!selectedWallet || !currentWallet) {
        return;
      }

      await signIn(selectedWallet, nearConfig.contractName);

      const currentWalletAccountId = currentWallet.getAccountId();
      const selectedWalletAccountId = selectedWallet.getAccountId();

      sendGAEvent({
        name: GA_EVENTS.SWITCH_WALLET,
        accountId: selectedWalletAccountId,
        params: {
          wallet: walletType,
          previousAccountId: currentWalletAccountId,
        },
      });

      if (currentWalletAccountId !== selectedWalletAccountId) {
        router.reload();
      }
    },
    [currentWallet, getWallet, nearConfig.contractName, router, signIn]
  );

  const walletContext = useMemo(() => {
    return {
      nearService: currentWallet ? new SputnikNearService(currentWallet) : null,
      availableWallets: availableWallets.map(availableWallet =>
        availableWallet.walletMeta()
      ),
      accountId: currentWallet?.getAccountId() ?? '',
      currentWallet: currentWallet?.getWalletType() ?? null,
      connectingToWallet,
      availableAccounts,
      pkAndSignature,
      login,
      logout,
      switchAccount,
      switchWallet,
    };
  }, [
    availableAccounts,
    availableWallets,
    connectingToWallet,
    currentWallet,
    login,
    logout,
    pkAndSignature,
    switchAccount,
    switchWallet,
  ]);

  return (
    <WalletContext.Provider value={walletContext}>
      {children}
      {connectingToWallet && (
        <ConnectingWalletModal
          isOpen
          onClose={() => toggleConnection(false)}
          walletType={walletContext?.currentWallet ?? WalletType.NEAR}
        />
      )}
    </WalletContext.Provider>
  );
};

export function useWalletContext(): WalletContext {
  return useContext(WalletContext);
}
