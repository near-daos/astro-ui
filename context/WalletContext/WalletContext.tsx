import {
  FC,
  useMemo,
  useState,
  useEffect,
  useContext,
  useCallback,
  createContext,
} from 'react';
import get from 'lodash/get';
import { useBoolean } from 'react-use';
import { useRouter } from 'next/router';

import { WalletType } from 'types/config';
import {
  WalletMeta,
  WalletService,
} from 'services/sputnik/SputnikNearService/walletServices/types';

import { configService } from 'services/ConfigService';
import { CookieService } from 'services/CookieService';
import { ACCOUNT_COOKIE } from 'constants/cookies';
import { SputnikNearService } from 'services/sputnik';
import { ConnectingWalletModal } from 'astro_2.0/features/Auth/components/ConnectingWalletModal';

import { GA_EVENTS, sendGAEvent } from 'utils/ga';
import { isSelectorWalletType } from 'utils/isSelectorWalletType';

import {
  NEAR_WALLET_METADATA,
  SENDER_WALLET_METADATA,
} from 'services/sputnik/SputnikNearService/walletServices/constants';

import { PkAndSignature } from './types';

import { useWallet } from './hooks/useWallet';
import { usePkAndSignature } from './hooks/usePkAndSignature';
import { useAvailableAccounts } from './hooks/useAvailableAccounts';
import { useSelectorLsAccount } from './hooks/walletSelector/useSelectorLsAccount';

export interface WalletContext {
  availableWallets: WalletMeta[];
  currentWallet: WalletType | null;
  accountId: string;
  connectingToWallet: boolean;
  login: (walletType: WalletType) => Promise<void>;
  logout: () => Promise<void>;
  switchAccount: (walletType: WalletType, accountId: string) => void;
  switchWallet: (walletType: WalletType) => void;
  availableAccounts: string[];
  pkAndSignature: PkAndSignature | null;
  // todo get rid of
  nearService: SputnikNearService | null;
}

const WalletContext = createContext<WalletContext>({} as WalletContext);

export const WrappedWalletContext: FC = ({ children }) => {
  const { nearConfig } = configService.get();

  const router = useRouter();

  const [, , removeSelectorLsAccount] = useSelectorLsAccount();
  const [connectingToWallet, setConnectingToWallet] = useBoolean(false);

  const { getWallet, setWallet, currentWallet, removePersistedWallet } =
    useWallet();

  const pkAndSignature = usePkAndSignature(currentWallet);
  const availableAccounts = useAvailableAccounts(currentWallet);
  const [currentAccount, setCurrentAccount] = useState('');

  useEffect(() => {
    async function updateAccountId() {
      if (currentWallet) {
        const account = await currentWallet.getAccountId();

        setCurrentAccount(account);
      }
    }

    updateAccountId();
  }, [currentWallet]);

  const signIn = useCallback(
    async (wallet: WalletService, contractName: string) => {
      setConnectingToWallet(true);

      try {
        await wallet.signIn(contractName);

        setWallet(wallet);

        setConnectingToWallet(false);

        const accountId = await wallet.getAccountId();

        sendGAEvent({
          name: GA_EVENTS.SIGN_IN,
          accountId,
        });
      } catch (e) {
        setConnectingToWallet(false);
      }
    },
    [setWallet, setConnectingToWallet]
  );

  const login = useCallback(
    async (walletType: WalletType) => {
      const wallet = await getWallet(walletType);

      if (!wallet) {
        return;
      }

      await signIn(wallet, nearConfig.contractName);
    },
    [signIn, getWallet, nearConfig.contractName]
  );

  const logout = useCallback(async () => {
    const accountId = (await currentWallet?.getAccountId()) ?? '';

    sendGAEvent({
      name: GA_EVENTS.SIGN_OUT,
      accountId,
    });

    CookieService.remove(ACCOUNT_COOKIE, { path: '/' });
    removePersistedWallet();

    await currentWallet?.logout();

    router.reload();
  }, [currentWallet, removePersistedWallet, router]);

  const switchAccount = useCallback(
    async (walletType: WalletType, accountId: string) => {
      // Currently its NEAR wallet specific only
      if (walletType !== WalletType.NEAR) {
        return;
      }

      const nearWallet = await getWallet(WalletType.NEAR);

      if (!nearWallet) {
        return;
      }

      // eslint-disable-next-line no-underscore-dangle
      const keypair = await nearWallet
        .getKeyStore()
        .getKey(nearConfig.networkId, accountId);

      if (!keypair) {
        return;
      }

      const authData = {
        accountId,
        allKeys: [keypair.getPublicKey().toString()],
      };

      // new wallet instance will take the new auth_key and will reinit the account
      window.localStorage.setItem(
        'sputnik_wallet_auth_key',
        JSON.stringify(authData)
      );

      CookieService.set(ACCOUNT_COOKIE, accountId, {
        path: '/',
      });

      sendGAEvent({
        name: GA_EVENTS.SWITCH_ACCOUNT,
        accountId,
      });

      router.reload();
    },
    [getWallet, nearConfig.networkId, router]
  );

  const switchWallet = useCallback(
    async (walletType: WalletType) => {
      const selectedWallet = await getWallet(walletType);

      if (!selectedWallet || !currentWallet) {
        return;
      }

      await signIn(selectedWallet, nearConfig.contractName);

      const currentWalletAccountId = await currentWallet.getAccountId();
      const selectedWalletAccountId = await selectedWallet.getAccountId();

      sendGAEvent({
        name: GA_EVENTS.SWITCH_WALLET,
        accountId: selectedWalletAccountId,
        params: {
          wallet: walletType,
          previousAccountId: currentWalletAccountId,
        },
      });

      if (currentWalletAccountId !== selectedWalletAccountId) {
        if (isSelectorWalletType(currentWallet.getWalletType())) {
          removeSelectorLsAccount();
        }

        router.reload();
      }
    },
    [
      router,
      signIn,
      getWallet,
      currentWallet,
      nearConfig.contractName,
      removeSelectorLsAccount,
    ]
  );

  const walletContext = useMemo(() => {
    const availableWallets = [NEAR_WALLET_METADATA];

    const senderWalletAvailable = get(window, 'near.isSender') || false;

    if (senderWalletAvailable) {
      availableWallets.push(SENDER_WALLET_METADATA);
    }

    return {
      login,
      logout,
      switchWallet,
      switchAccount,
      pkAndSignature,
      availableWallets,
      availableAccounts,
      connectingToWallet,
      accountId: currentAccount,
      currentWallet: currentWallet?.getWalletType() ?? null,
      nearService: currentWallet ? new SputnikNearService(currentWallet) : null,
    };
  }, [
    login,
    logout,
    switchWallet,
    switchAccount,
    currentWallet,
    pkAndSignature,
    currentAccount,
    availableAccounts,
    connectingToWallet,
  ]);

  return (
    <WalletContext.Provider value={walletContext}>
      {children}
      {connectingToWallet && (
        <ConnectingWalletModal
          isOpen
          onClose={() => setConnectingToWallet(false)}
          walletType={walletContext?.currentWallet ?? WalletType.NEAR}
        />
      )}
    </WalletContext.Provider>
  );
};

export function useWalletContext(): WalletContext {
  return useContext(WalletContext);
}
