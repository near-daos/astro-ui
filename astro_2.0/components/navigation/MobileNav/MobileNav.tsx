import React, { VFC } from 'react';

import { useWalletContext } from 'context/WalletContext';

import {
  CFC_LIBRARY_NAV_CONFIG,
  MY_DAOS_NAV_CONFIG,
  MY_FEED_NAV_CONFIG,
  ASTRO_FEED_NAV_CONFIG,
  DISCOVER_NAV_CONFIG,
} from 'astro_2.0/components/navigation/navConfig';
import { NavButton } from 'astro_2.0/components/navigation/NavButton';
import { useDaoIds } from 'hooks/useDaoIds';

import styles from './MobileNav.module.scss';

export const MobileNav: VFC = () => {
  const { accountId } = useWalletContext();
  const myDaosIds = useDaoIds(accountId);

  const navConfig = accountId
    ? [
        MY_FEED_NAV_CONFIG,
        MY_DAOS_NAV_CONFIG,
        CFC_LIBRARY_NAV_CONFIG,
        DISCOVER_NAV_CONFIG,
        ASTRO_FEED_NAV_CONFIG,
      ]
    : [DISCOVER_NAV_CONFIG, CFC_LIBRARY_NAV_CONFIG, ASTRO_FEED_NAV_CONFIG];

  const navItems = navConfig.map(conf => {
    return (
      <NavButton {...conf} mobile myDaosIds={myDaosIds} key={conf.label} />
    );
  });

  return <div className={styles.root}>{navItems}</div>;
};
