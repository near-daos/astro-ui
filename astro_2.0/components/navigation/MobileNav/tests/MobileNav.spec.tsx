/* eslint-disable @typescript-eslint/ban-ts-comment */
import { render } from 'jest/testUtils';

import { MobileNav } from 'astro_2.0/components/navigation/MobileNav';
import { useWalletContext } from 'context/WalletContext';

jest.mock('context/WalletContext', () => {
  return {
    useWalletContext: jest.fn(),
  };
});

jest.mock('astro_2.0/components/navigation/NavButton', () => {
  return {
    NavButton: ({ label }: { label: string }) => (
      <div data-testid="nav-item">{label}</div>
    ),
  };
});

describe('Mobile nav', () => {
  it('Should render partial navigation if user did not log in', () => {
    // @ts-ignore
    useWalletContext.mockImplementation(() => ({
      accountId: undefined,
    }));

    const { container, getAllByTestId } = render(<MobileNav />);

    expect(container).toMatchSnapshot();
    expect(getAllByTestId('nav-item')).toHaveLength(3);
  });

  it('Should render full navigation for logged user', () => {
    // @ts-ignore
    useWalletContext.mockImplementation(() => ({
      accountId: 'accountId',
    }));

    const { container, getAllByTestId } = render(<MobileNav />);

    expect(container).toMatchSnapshot();
    expect(getAllByTestId('nav-item')).toHaveLength(5);
  });
});
