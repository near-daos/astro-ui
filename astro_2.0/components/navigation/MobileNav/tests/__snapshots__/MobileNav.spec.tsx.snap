// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mobile nav Should render full navigation for logged user 1`] = `
<div>
  <div
    class="root"
  >
    <div
      data-testid="nav-item"
    >
      My Feed
    </div>
    <div
      data-testid="nav-item"
    >
        My DAOs
    </div>
    <div
      data-testid="nav-item"
    >
      Actions Library
    </div>
    <div
      data-testid="nav-item"
    >
      Discovery
    </div>
    <div
      data-testid="nav-item"
    >
      Global Feed
    </div>
  </div>
</div>
`;

exports[`Mobile nav Should render partial navigation if user did not log in 1`] = `
<div>
  <div
    class="root"
  >
    <div
      data-testid="nav-item"
    >
      Discovery
    </div>
    <div
      data-testid="nav-item"
    >
      Actions Library
    </div>
    <div
      data-testid="nav-item"
    >
      Global Feed
    </div>
  </div>
</div>
`;
